tests/functional/test_plugins.py

```diff
@@ -0,0 +1,173 @@
+import sys
+
+import pytest
+
+from botocore import utils
+from botocore.session import get_session
+from tests import mock
+
+
+class TestPluginModule:
+    """A mock plugin module for testing client plugin loading."""
+
+    def __init__(self):
+        self.events_seen = []
+
+    def initialize_client_plugin(self, client):
+        client.meta.events.register('before-call', self.increment_calls)
+
+    def increment_calls(self, **kwargs):
+        self.events_seen.append(kwargs)
+
+
+class RecursivePluginModule:
+    def initialize_client_plugin(self, client):
+        client.meta.events.register('before-call.s3.*', self.create_client)
+
+    def create_client(self, **kwargs):
+        session = get_session()
+        client = utils.create_nested_client(
+            session, "s3", region_name="us-west-2"
+        )
+        client.list_buckets()
+
+
+@pytest.fixture
+def test_plugin():
+    """Fixture to create and register test plugin module in sys.modules."""
+    module = TestPluginModule()
+    sys.modules['test_plugin'] = module
+    yield module
+    sys.modules.pop('test_plugin', None)
```

Comment:
Do we have another way to register this without having to patch sys.modules? It seems like we could accidentally break other libraries.

Reply:
If we are testing loading arbitrary modules, it needs to be exposed to the interpreter in one form or another.  I've modified this to be appending the sys path instead now, but am not sure if there's a better way.